server/logging/firehose.go

```diff
@@ -0,0 +1,162 @@
+	}
+	client := firehose.New(sess)
+
+	f := &firehoseLogWriter{
+		client: client,
+		stream: stream,
+		logger: logger,
+	}
+	if err := f.validateStream(); err != nil {
+		return nil, errors.Wrap(err, "create Firehose writer")
+	}
+	return f, nil
+}
+
+func (f *firehoseLogWriter) validateStream() error {
+	out, err := f.client.DescribeDeliveryStream(
+		&firehose.DescribeDeliveryStreamInput{
+			DeliveryStreamName: &f.stream,
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "describe stream %s", f.stream)
+	}
+
+	if (*(*out.DeliveryStreamDescription).DeliveryStreamStatus) != firehose.DeliveryStreamStatusActive {
+		return errors.Errorf("delivery stream %s not active", f.stream)
+	}
+
+	return nil
+}
+
+func (f *firehoseLogWriter) Write(logs []json.RawMessage) error {
+	var records []*firehose.Record
+	totalBytes := 0
+	for _, log := range logs {
+		// We don't really have a good option for what to do with logs
+		// that are too big for Firehose. This behavior is consistent
+		// with osquery's behavior in the Firehose logger plugin, and
+		// the beginning bytes of the log should help the Fleet admin
+		// diagnose the query generating huge results.
+		if len(log) > firehoseMaxSizeOfRecord {
+			level.Info(f.logger).Log(
+				"msg", "dropping log over 1MB Firehose limit",
+				"size", len(log),
+				"log", string(log[:100])+"...",
+			)
+			continue
+		}
+
+		// If adding this log will exceed the limit on number of
+		// records in the batch, or the limit on total size of the
+		// records in the batch, we need to push this batch before
+		// adding any more.
+		if len(records) >= firehoseMaxRecordsInBatch ||
+			totalBytes+len(log) >= firehoseMaxSizeOfBatch {
+			if err := f.putRecordBatch(0, records); err != nil {
+				return errors.Wrap(err, "put records")
+			}
+			totalBytes = 0
+			records = nil
+		}
+
+		records = append(records, &firehose.Record{Data: []byte(log)})
+		totalBytes += len(log)
+	}
+
+	// Push the final batch
+	if len(records) > 0 {
+		if err := f.putRecordBatch(0, records); err != nil {
+			return errors.Wrap(err, "put records")
+		}
+	}
+
+	return nil
+}
+
+func (f *firehoseLogWriter) putRecordBatch(try int, records []*firehose.Record) error {
+	if try > 0 {
+		time.Sleep(100 * time.Millisecond * time.Duration(math.Pow(2.0, float64(try))))
+	}
+	input := &firehose.PutRecordBatchInput{
+		DeliveryStreamName: &f.stream,
+		Records:            records,
+	}
+
+	output, err := f.client.PutRecordBatch(input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == firehose.ErrCodeServiceUnavailableException && try < firehoseMaxRetries {
+				// Retry with backoff
+				return f.putRecordBatch(try+1, records)
+			}
+		}
+
+		// Not retryable or retries expired
+		return err
+	}
+
+	// Check errors on individual records
+	if output.FailedPutCount != nil && *output.FailedPutCount > 0 {
+		if try >= firehoseMaxRetries {
+			return errors.Errorf(
+				"failed to put %d records, retries exhausted",
```

Comment:
Kind of a shame we can't use `output.RequestResponses` to give the more info on why N records couldn't be pushed. I didn't see it, but is there any point where we make error status codes available to a user?

If we don't, do you think we should?

Reply:
It's not really clear the best way to expose the potentially up to 500 errors to the user. Maybe the error from a single record would be representative and we could expose that?

Reply:
I exposed the first error.